src/core/shutdown.c

```diff
@@ -159,6 +163,78 @@ static int switch_root_initramfs(void) {
         return switch_root("/run/initramfs", "/oldroot", false, MS_BIND);
 }
 
+static bool sync_making_progress(unsigned long long *prev_dirty) {
+        _cleanup_free_ char *dirty_string = NULL;
+        unsigned long long dirty = 0;
+        int r;
+
+        r = get_proc_field("/proc/meminfo", "Dirty", WHITESPACE, &dirty_string);
+        if (r < 0) {
+                log_error_errno(r, "Failed to retrieve Dirty from /proc/meminfo: %m");
+                return false;
+        }
+
+        r = safe_atollu(dirty_string, &dirty);
+        if (r < 0) {
+                log_error_errno(r, "Failed to parse Dirty from /proc/meminfo: %s: %m",
+                                dirty_string);
+                return false;
+        }
+
+        if (*prev_dirty > dirty)
+                r = true;
+        else
+                r = false;
+
+        *prev_dirty = dirty;
+
+        return r;
+}
+
+static void sync_with_progress(void) {
+        unsigned checks;
+        pid_t pid;
+        int r;
+        unsigned long long dirty = -1;
```

Comment:
even though C allows this, this is kinda ugly: you make it unsigned, but then assign a negative value to it. I think it would be much nicer if this would be explicitly cast, so that this fact is explicit.

Reply:
+1.  ULONG_LONG_MAX is also a possibility.